### DISCUSSION TITLE
Automatic IT generation

### DISCUSSION THREAD

**Discussion by mmussini** [OAI/Arazzo-Specification#213] [General]
Hi guys,
very nice project.
Are you aware of any tool that integrates Arazzo to automatically generate Integration Tests (e.g. newman)?

Thanks in advance for any hint,
Manuel

**Comment by microapiframework** [comment]
Hi. Nothing yet. It was just released so it may be some time before any tooling is readily available.

**Reply by mmussini** [reply]
Thanks!

**Reply by orubel** [reply]
It wouldn't work anyway as currently it only supports external redirection with single threaded architectures. This is not even close to a one size fits all solution and has no way to address internal redirection in the app (see forward, step functions, etc)

I brought this up and the lead was so ill-informaed as to accuse me of making of the term 'internal redirection' as well as industry standards such as 'forward' and 'step functions' and 'loopback'

**Reply by leidenheit** [reply]
From my point of view, Arazzo can be considered as a typical client for whom the details of redirects and forwards (whether sync or async) are not relevant. These processes should be transparent to the client. The most important aspect for the client is ensuring that responses are correct and accurate.

**Reply by orubel** [reply]
@leidenheit It should be transparent but if you are doing call flows, you will need to distinguish what type of redirect you are using since there are TWO TYPES.

**Comment by leidenheit** [comment]
Hi @mmussini I'm working on implementing this, and if you're interested, please feel free to check out Itarazzo:
- [Itarazzo-Client](https://github.com/leidenheit/itarazzo-client/tree/develop) -  A Dockerized version designed to run Arazzo specifications effortlessly.
- [Itarazzo-Library](https://github.com/leidenheit/itarazzo-library) - Handles parsing, validation, and execution of Arazzo specifications.

**Please Note**: It's still in an early stage, so there are some bugs.

**Reply by orubel** [reply]
Based on your statement, it either doesnt use redirection ... or uses only one kind of redirection.

**Reply by leidenheit** [reply]
Itarazzo does not contain any special behavior regarding redirects at all. As mentioned, it operates at a higher level and interacts with server APIs as a black box as specified in the Arazzo Specification. It seems I have a gap in understanding the point you are highlighting, and I would like to learn more about it.

**Reply by orubel** [reply]
@leidenheit the highest level in an api application is the 'front controller'(ie requestDispatcher for java) as it handles the request/response. Thats where half of redirection occurs. What other 'higher level' are you talking about? The SERVLET??? It doesnt handle the request/response except to hand off to requestDispatcher

**Reply by leidenheit** [reply]
It seems like there might be some frustration, and I want to gently remind you to stay focused on an objective discussion. I recognize that you might have more experience in this context than I do, so let me explain my understanding differently:

Itarazzo uses RestAssured for OAS-related communication. RestAssured is a test framework for RESTful web services. By default, RestAssured automatically follows all types of redirects, which are 3xx status codes. From my perspective, this should be sufficient.

Do you agree?

**Reply by orubel** [reply]
@leidenheit No because it doesn't. I actaully pointed this out to the creators of RestAssurred. It only follows 'external redirects'. There are two types of redirects and ALL modern web servers support them. And **regardless of what Darrel Miller and Henry Andrews may think**, they exist and Roy Fielding confirmed they have since the very first HTTPD server.

**Reply by leidenheit** [reply]
That's interesting. How did they handled your discovery?

**Reply by orubel** [reply]
Apathy and ignorance (much like openapi #261). Because they are a product (not a standard), they chose their market... tools using reactor pattern. Those that use their tool while using a framework with 'proactor pattern' tend to suffer security issues as a result of them not addressing these call flows.

**Reply by leidenheit** [reply]
Thank you for the information. I believe I now understand your point. Referring back to Arazzo, your contributions seem to suggest introducing a property that defines how a potential request redirect should be managed - either synchronously or asynchronously. Based on this property, the tools would then need to behave accordingly. Is that correct?

**Reply by orubel** [reply]
In proactor vs reactor mainly (not necessarily sync vs async - see AWS Step Functions or nodejs loopback plugin; they are call flow add-ons that act like internal redirects). I mean both are technically synchronous its just one is synchronous in the app and one is synchronous at the app server ;)

My point is that api frameworks are NOT merely reactor pattern and as such can redirect both ways(internally and externally) and they should plan for it otherwise they risk security holes. According to Roy Fielding, this has been built-in and planned functionality since the very first HTTPD server

Alot of reactor frameworks cannot do internal redirects; They gimped themselves to get quick speed up front (by moving front controller out to app server) but bad concurrency as a result and no internal redirection.

Regardless, a STANDARD needs to work for ALL api implementations... not just 50% of them (and then expect the rest to understand or know that this does not cover their asses)

**Comment by emretezisci** [comment]
Hi @mmussini, (full disclosure: I work at Speakeasy)

We’ve integrated Arazzo into Speakeasy so you can automatically generate multi-step integration tests—kind of like a Newman experience, but straight from your OpenAPI specs. It also comes with test generation, data generation, a built-in mock server, and more. If you want to learn more or need help getting started, just let me know! (https://www.speakeasy.com/post/release-contract-testing)

**Reply by mmussini** [reply]
Thank you very much for the pointers. I'll look at it asap.
